src/Evolution/Systems/Cce/WorldtubeInterfaceManager.hpp

```diff
@@ -0,0 +1,144 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/optional.hpp>
+#include <deque>
+#include <memory>
+#include <tuple>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Time/TimeStepId.hpp"
+
+namespace Cce {
+/// \cond
+class GHLockstepInterfaceManager;
+/// \endcond
+
+/*!
+ * \brief Abstract base class for storage and retrieval of generalized harmonic
+ * quantities communicated from a Cauchy simulation to the Cce system.
+ *
+ * \details The functions that are required to be overriden in the derived
+ * classes are:
+ * - `GHWorldtubeInterfaceManager::get_clone()`: should return a
+ * `std::unique_ptr<GHWorldtubeInterfaceManager>` with cloned state.
+ * - `GHWorldtubeInterfaceManager::insert_gh_data()`: should store the portions
+ * of the provided generalized harmonic data that are required to provide useful
+ * boundary values for the CCE evolution at requested timesteps.
+ * - `GHWorldtubeInterfaceManager::request_gh_data()`: should register requests
+ * from the CCE evolution for boundary data.
+ * - `GHWorldtubeInterfaceManager::try_retrieve_gh_data()`: should return a
+ * `std::tuple<TimeStepId, tnsr::aa<DataVector, 3>,
+ *  tnsr::iaa<DataVector, 3>, tnsr::aa<DataVector, 3>>` containing the boundary
+ * data associated with the least recent requested timestep if enough data has
+ * been supplied via `insert_gh_data()` to determine the boundary data.
+ * Otherwise, return a `boost::none` to indicate that the CCE system must
+ * continue waiting for generalized harmonic input.
+ */
+class GHWorldtubeInterfaceManager : public PUP::able {
+ public:
+  using creatable_classes = tmpl::list<GHLockstepInterfaceManager>;
+
+  WRAPPED_PUPable_abstract(GHWorldtubeInterfaceManager);  // NOLINT
+
+  virtual std::unique_ptr<GHWorldtubeInterfaceManager> get_clone() const
+      noexcept = 0;
+
+  virtual void insert_gh_data(
+      const TimeStepId& time_id,
+      const tnsr::aa<DataVector, 3>& spacetime_metric,
+      const tnsr::iaa<DataVector, 3>& phi, const tnsr::aa<DataVector, 3>& pi,
+      const tnsr::aa<DataVector, 3>& dt_spacetime_metric,
+      const tnsr::iaa<DataVector, 3>& dt_phi,
+      const tnsr::aa<DataVector, 3>& dt_pi) noexcept = 0;
+
+  virtual void request_gh_data(const TimeStepId&) noexcept = 0;
+
+  virtual auto try_retrieve_first_ready_gh_data() noexcept -> boost::optional<
```

Comment:
can this be a const function?

Reply:
the retrieval `pop`s data from the deque, so I think it shouldn't be const (and could only be if we made the `deque` mutable).

Reply:
Ah I see, that should probably be clearer from the function name or documentation. I'm not coming up with anything obvious for the name, `retrieve_and_remove_first_ready_gh_data()`? I think the `optional` mostly covers the `try` part :man_shrugging: